packages/firestore/scripts/prune-dts.ts

```diff
@@ -0,0 +1,339 @@
+function isExported(modifiers?: ts.ModifiersArray): boolean {
+  return !!modifiers?.find(m => m.kind === ts.SyntaxKind.ExportKeyword);
+}
+
+/**
+ * Replaces an existing constructor implementation if the constructor is marked
+ * with the JSDod tag `@hideconstructor`. The replaced constructor can either
+ * have `private` visibility` or `proctected`. To generate a protected
+ * constructor, specify `@hideconstructor proctected`.
+ *
+ * Returns either the modified constructor or the existing constructor if no
+ * modification was needed.
+ */
+function maybeHideConstructor(node: ts.ConstructorDeclaration) {
+  const hideConstructorTag = ts
+    .getJSDocTags(node)
+    ?.find(t => t.tagName.escapedText === 'hideconstructor');
+
+  if (hideConstructorTag) {
+    const modifier = ts.createModifier(
+      hideConstructorTag.comment === 'protected'
+        ? ts.SyntaxKind.ProtectedKeyword
+        : ts.SyntaxKind.PrivateKeyword
+    );
+    return ts.createConstructor(
+      node.decorators,
+      [modifier],
+      /*parameters=*/ [],
+      /* body= */ undefined
+    );
+  } else {
+    return node;
+  }
+}
+
+/**
+ * Examines `extends` and `implements` clauses and removes or replaces them if
+ * they refer to a non-exported type. When an export is removed, all members
+ * from the removed class are merged into the provided class or interface
+ * declaration.
+ *
+ * @example
+ * Input:
+ * class Foo {
+ *   foo: string;
+ * }
+ * export class Bar extends Foo {}
+ *
+ * Output:
+ * export class Bar {
+ *   foo: string;
+ * }
+ */
+function prunePrivateImports<
+  T extends ts.InterfaceDeclaration | ts.ClassDeclaration
+>(typeChecker: ts.TypeChecker, sourceFile: ts.SourceFile, node: T): T {
+  const currentType = typeChecker.getTypeAtLocation(node);
+  const currentName = currentType?.symbol?.name;
+  const currentMembers = currentType?.symbol?.members;
+
+  // The list of heritage clauses after all private symbols are removed.
+  const prunedHeritageClauses: ts.HeritageClause[] = [];
+  // Additional members that are copied from the private symbols into the public
+  // symbols
+  const additionalMembers: ts.NamedDeclaration[] = [];
+
+  for (const heritageClause of node.heritageClauses || []) {
+    const exportedTypes: ts.ExpressionWithTypeArguments[] = [];
+    for (const type of heritageClause.types) {
+      if (isExported(type.modifiers)) {
+        exportedTypes.push(type);
+      } else {
+        const publicName = extractPublicName(
+          typeChecker,
+          sourceFile,
+          type.expression
+        );
+        if (publicName && publicName != currentName) {
+          // If there is a public type that we can refer to, update the import
+          // statement to refer to the public type.
+          exportedTypes.push(
+            ts.updateExpressionWithTypeArguments(
+              type,
+              type.typeArguments,
+              ts.createIdentifier(publicName)
```

Comment:
IIUC, this can potentially make the class/interface has more members than it really has. Considering this:
```
class A {
   data: string;
}

export class C extends A {
   additionalData: number;
}

export class B  extends A {
}

```

 will become 

```
export class C {
  data: string;
  additionalData: number;
}

export class B  extends C {
}
```

Now `B` has `additionalData` it didn't have before.

I think the safest strategy is always copy the members from the private symbols instead of replacing it with a public one that extends it.

Reply:
I'm confused. Is B2 B and C2 C from my example? Or are you giving a different example?